dll/win32/kernel32/client/appcache.c

```diff
@@ -139,20 +139,122 @@ IsShimInfrastructureDisabled(VOID)
  * @unimplemented
  */
 BOOL
-WINAPI
-BaseCheckAppcompatCache(IN PWCHAR ApplicationName,
-                        IN HANDLE FileHandle,
-                        IN PWCHAR Environment,
-                        OUT PULONG Reason)
+BasepShimCacheCheckBypass(
+    _In_ PCWSTR ApplicationName,
+    _In_ HANDLE FileHandle,
+    _In_opt_ PCWSTR Environment,
+    _In_ BOOL bUnknown,
+    _Out_opt_ PULONG pdwReason)
 {
-    DPRINT("BaseCheckAppcompatCache is UNIMPLEMENTED\n");
+    DPRINT("fixme:(%S, %p, %S, %d, %p)\n", ApplicationName, FileHandle, Environment, bUnknown,
+           pdwReason);
+    return FALSE;
+}
 
-    if (Reason) *Reason = 0;
+/*
+ * @implemented
+ */
+BOOL
+BasepShimCacheSearch(
+    _In_ PCWSTR ApplicationName,
+    _In_ HANDLE FileHandle)
+{
+    APPHELP_CACHE_SERVICE_LOOKUP Lookup;
+    RtlInitUnicodeString(&Lookup.ImageName, ApplicationName);
+    Lookup.ImageHandle = FileHandle;
+    return NT_SUCCESS(NtApphelpCacheControl(ApphelpCacheServiceLookup, &Lookup));
+}
 
-    // We don't know this app.
+/*
+ * @unimplemented
+ */
+BOOL
+BasepCheckCacheExcludeList(
+    _In_ PCWSTR ApplicationName)
+{
     return FALSE;
 }
 
+/*
+ * @unimplemented
+ */
+BOOL
+BasepCheckCacheExcludeCustom(
+    _In_ PCWSTR ApplicationName)
+{
+    return FALSE;
+}
+
+/*
+ * @implemented
+ */
+VOID
+BasepShimCacheRemoveEntry(
+    _In_ PCWSTR ApplicationName)
+{
+    APPHELP_CACHE_SERVICE_LOOKUP Lookup;
+    RtlInitUnicodeString(&Lookup.ImageName, ApplicationName);
+    Lookup.ImageHandle = INVALID_HANDLE_VALUE;
+    NtApphelpCacheControl(ApphelpCacheServiceRemove, &Lookup);
+}
+
+/*
+ * @unimplemented
+ */
+BOOL
+BasepShimCacheLookup(
+    _In_ PCWSTR ApplicationName,
+    _In_ HANDLE FileHandle)
+{
+    DPRINT("fixme:(%S, %p)\n", ApplicationName, FileHandle);
+
+    if (!BasepShimCacheSearch(ApplicationName, FileHandle))
+        return FALSE;
+
+    if (!BasepCheckCacheExcludeList(ApplicationName) ||
+        !BasepCheckCacheExcludeCustom(ApplicationName))
+    {
+        BasepShimCacheRemoveEntry(ApplicationName);
+        return FALSE;
+    }
+
+    return TRUE;
+}
+
+/*
+ * @implemented
+ */
+BOOL
+WINAPI
+BaseCheckAppcompatCache(
+    _In_ PCWSTR ApplicationName,
+    _In_ HANDLE FileHandle,
+    _In_opt_ PCWSTR Environment,
+    _Out_opt_ PULONG pdwReason)
+{
+    BOOL ret = FALSE;
+    ULONG dwReason;
+
+    DPRINT("(%S, %p, %S, %p)\n", ApplicationName, FileHandle, Environment, pdwReason);
+
+    dwReason = 0;
+    if (BasepShimCacheCheckBypass(ApplicationName, FileHandle, Environment, TRUE, &dwReason))
+    {
+        dwReason |= 2;
```

Comment:
It's unknown.

Reply:
or is there some software that triggers this?

Reply:
Magic behavior like this is not what I would consider good practice - it gives the impression we're mindlessly reverse-engineering. I would recommend one of the following approaches:
* Ideally, include a test that covers this behavior. That explains how you found the information, how others can verify it, lets us spot regressions etc -- best solution by far. Hopefully that would lead to a good-enough understanding of the code to choose names for these flags, but if not, you can at least add a comment that describes when/why the flag is needed.
* If there is external software that relies on this (could be a Windows dll I guess), mention this in the commit description and/or a comment. This way it's clear why you're adding it, what has been used to determine the values and can be used to test changes etc.
* If neither of the above is possible, invest the time to understand it and write good code based on your understanding. That would include defining reasonably-named constants for flag values.